packages/server/src/trpc-routers/__tests_e2e__/swap-router.test.ts

```diff
@@ -0,0 +1,1160 @@
+    | undefined;
+  for (const split of reply.split) {
+    expect(split.initialAmount).toBeInstanceOf(Int);
+
+    expect(Array.isArray(split.pools)).toBeTruthy();
+    expect(split.pools.length).toBeGreaterThan(0);
+
+    expect(split.tokenInDenom).toBe(astroMinimalDenom);
+
+    pclPool = split.pools.find(
+      (pool) => pool.type === "cosmwasm-astroport-pcl"
+    );
+
+    expect(Array.isArray(split.tokenOutDenoms)).toBeTruthy();
+    expect(split.tokenOutDenoms.length).toBeGreaterThan(0);
+  }
+
+  // Sum of all split amount should equal token in amount
+  const splitAmountSum = reply.split.reduce(
+    (acc, split) => acc.add(split.initialAmount),
+    new Int(0)
+  );
+  expect(splitAmountSum.equals(new Int(tokenInAmount))).toBeTruthy();
+
+  // PCL assets should go through a pcl pool
+  expect(pclPool).toBeDefined();
+  expect(pclPool!.inCurrency).toEqual(makeMinimalAsset(astroAsset.rawAsset));
+  expect(pclPool!.outCurrency).toEqual(makeMinimalAsset(osmoAsset.rawAsset));
+
+  // name
+  expect(reply.name).toBe("sidecar");
+
+  // timeMs
+  expect(isNumeric(reply.timeMs)).toBeTruthy();
+
+  // Token in fee amount fiat value
+  expect(reply.tokenInFeeAmountFiatValue).toBeInstanceOf(PricePretty);
+
+  const tokenInFeeAmountFiatValue = reply
+    .tokenInFeeAmountFiatValue!.toDec()
+    .toString();
+  expect(isNumeric(tokenInFeeAmountFiatValue)).toBeTruthy();
+  expect(parseFloat(tokenInFeeAmountFiatValue)).toBeGreaterThan(0);
+
+  // token out price
+  expect(reply.tokenOutPrice).toBeInstanceOf(PricePretty);
+
+  const tokenOutPrice = reply.tokenOutPrice!.toDec().toString();
+  expect(isNumeric(tokenOutPrice)).toBeTruthy();
+  expect(parseFloat(tokenOutPrice)).toBeGreaterThan(0);
+
+  // amount fiat value
+  expect(reply.amountFiatValue).toBeInstanceOf(PricePretty);
+
+  const amountFiatValue = reply.amountFiatValue!.toDec().toString();
+  expect(isNumeric(amountFiatValue)).toBeTruthy();
+  expect(parseFloat(amountFiatValue)).toBeGreaterThan(0);
+});
+
+it("TFM - ATOM <> OSMO - should return valid partial quote (no swap fee)", async () => {
+  const tokenInAmount = "1000000";
+  const reply = await caller.swapRouter.routeTokenOutGivenIn({
+    tokenInDenom: atomMinimalDenom,
+    tokenInAmount,
+    tokenOutDenom: osmoMinimalDenom,
+    preferredRouter: "tfm",
+  });
+
+  // Amount
+  expect(reply.amount).toBeInstanceOf(CoinPretty);
+  expect(reply.amount.currency).toEqual(makeMinimalAsset(osmoAsset.rawAsset));
+
+  const amount = reply.amount.toDec().toString();
+  expect(isNumeric(amount)).toBeTruthy();
+  // Make sure amount is not negative
+  expect(parseFloat(amount)).toBeGreaterThan(0);
+
+  // Swap fee
+  // expect(reply.swapFee).toBeInstanceOf(RatePretty);
+  // // Should match with the format of "0.1%"
+  // expect(reply.swapFee?.toString()).toMatch(percentageRegex);
+
+  // const swapFee = reply.swapFee!.toDec().toString();
+  // expect(isNumeric(swapFee)).toBeTruthy();
+  // expect(parseFloat(swapFee)).toBeGreaterThan(0);
```

Comment:
Is this something that TFM doesn't return? If so, what is our plan around this? I propose making an issue, outlining all fields that are missing. Then, we will share it with them, requesting to add the missing fields to the API

Reply:
We are only missing the swap fee from their quote. I can reach out to them. Also, do you know the formula to compute the effective fee?

Reply:
Sounds great, yeah please let them know.

Here's the effective fee:
https://github.com/osmosis-labs/sqs/blob/664ad77accf906ea61cd88eff0294f1be6fca779/router/usecase/quote.go#L92

It's computed above